Atomic.md

```diff
@@ -0,0 +1,270 @@
+
+    -   An abandoned folder is in .temp that can be safely
+        overwritten/deleted when the enumeration is retried
+
+-   One of the SetBitInFileFlags calls fails in
+     InitializeEmptyPlaceholder
+
+    -   Folder is marked as in virtualization root but not empty.
+
+        -   A partial placeholder is in .temp that can be safely
+            overwritten/deleted when the enumeration is retried
+
+    -   Folder is marked as empty but not in virtualization root
+
+        -   A partial placeholder is in .temp that can be safely
+            overwritten/deleted when the enumeration is retried
+
+    -   The rename() to copy the completed placeholder from .temp into
+        the virtualization root fails
+
+        -   We can retry from this error since nothing ever appeared in
+            the virtualization root. The placeholder will be overwritten
+            as part of the retry
+
+WritePlaceholderFile:
+
+-   Fopen() fails to create a file
+
+    -   We can simply retry and have this succeed. Nothing was created
+        in .temp
+
+-   Ftruncate fails to expand the file to the correct size (or is never
+     reached because of a crash)
+
+    -   We can simply retry and have this succeed. The file written in
+        .temp will be overwritten/deleted and recreated
+
+-   Memcpy'ing the providerID xattr fails (or is never reached because
+     of a crash)
+
+    -   We have a partial placeholder file in .temp that will be
+        overwritten by a retry.
+
+-   Memcpy'ing the contentID xattr fails (or is never reached because of
+     a crash)
+
+    -   We have a partial placeholder file in .temp that will be
+        overwritten by a retry.
+
+-   We crash before reaching InitializeEmptyPlaceholder
+
+    -   We have a partially created placeholder written into .temp that
+        will be overwritten on retry
+
+-   One of the SetBitInFileFlags calls fails in
+     InitializeEmptyPlaceholder
+
+    -   Folder is marked as in virtualization root but not empty.
+
+        -   A partial placeholder is in .temp that can be safely
+            overwritten/deleted when the enumeration is retried
+
+    -   Folder is marked as empty but not in virtualization root
+
+        -   A partial placeholder is in .temp that can be safely
+            overwritten/deleted when the enumeration is retried
+
+-   Chmod() fails or is never reached due to a crash
+
+    -   We can retry and have this succeed as the placeholder hasn't
+        left .temp yet
+
+    -   The rename() to copy the completed placeholder from .temp into
+        the virtualization root fails
+
+        -   We can retry from this error since nothing ever appeared in
+            the virtualization root. The placeholder will be overwritten
+            as part of the retry
+
+All of the potential holes identified above are filled by this design,
+so I believe this to be our best bet to create placeholders in an atomic
+matter.
```

Comment:
Based on my reading of the design, it sounds like placeholders are created in `.temp` and then moved into their final location one at a time.  What happens if VFSForGit is killed in the middle of this process?  Can it resume where it left off in the next enumeration?  How will this work?

Reply:
That is the correct reading of the document, I'll explicitly call that out here. 

We will replace all placeholders that are present in this folder (the scenario I had in mind was VFSForGit crashes/is killed, on re-mounting, the customer uses this time to checkout a new branch before retrying their enumeration to ensure they get stamped with the right contentID) with new ones rather than trying to optimize. We will NOT touch files that don't belong to VFSForGit (if a user goes to that folder without the kext loaded and starts creating new files).